betaflight/betaflight
# update antigravity for 4.4

Update to anti-gravity for 4.4. 

Flight tested, thoroughly.  Works well.  Fxes oscillation issues on yaw and also from excessive / prolonged gain. Improves stick responsiveness and flight feel.  Adds tuning options.  Saves a bit of flash, removes old code.

- removes antigravity I boost on yaw, since many quads will wobble from even a small amount of I boost.
- removes the old Step mode, Smooth is much more consistent, Step has issues.
- allows strength of the P boost to be adjusted relative to the I boost, in case the user has P gain wobbles
- changes the default value to 80

IMPORTANT:  **The default value for anti_gravity_gain in CLI is now 80**.  This change in value is so that the number means something.  80 will add 8 times the user's default iTerm amount with the fastest throttle change, as if the user had transiently increased their iTerm setting by a factor of 8.  

NOTE 1: **The new default AntiGravity gain value is quite strong**.  It should NOT be increased without carefully observing for evidence of transient oscillation when hitting full throttle rapidly, or when chopping throttle.  

NOTE 2:  **The new default AntiGravity gain value may be TOO MUCH for highly tuned race builds where the pilot is very rapidly changing throttle!**  Reduce the gain if the default value causes transient oscillation.

NOTE 3: **The increase in P in this PR is, by default, about 1/4th that for I**.  This is slightly less than the previous P increase.  **On defaults, a fast throttle change will transiently increase P by about 2-2.5 times.**  The relative amount of P gain can be increased or decreased by modifying the `anti_gravity_p_gain`, where 100 means normal, zero means no P gain.

**HOW TO TEST / TUNE:**
- Fly cleanly straight forward at a reasonable pitch angle. 
- Throttle up fast then had chop fast with sticks centred.  Avoid adding unintentional yaw inputs
- Check the amount of pitch wobble and look out for transient oscillation on pitch or roll
- Repeat with higher or lower antigravity gain values
- Finally do an A-B comparison with it ON vs OFF.

**Formal Tuning:**

- Set `anti_gravity_p_gain` to 0, so there is no P boost, only an iTerm increase.
- Enable the debug `DEBUG_ANTI_GRAVITY` and log flights
- Tune the I element first.  Adjust the `anti_gravity_gain` upwards in steps, eg 1 - 40 - 80 - 120.  Any reasonable steps will do, but start at 1 (effectively none) to get a baseline.  As you move up, check for improvements in the pitch wobble when rapidly cutting throttle.  The wobble will be attenuated as gain increases. Too much gain will case iTerm oscillations with rapid throttle changes, both on throttle up and throttle down (check the log).  Choose the value that gives the best result, keeping below the value that starts to cause transient iTerm wobbles.  It's normal to not fully control the pitch wobble with I alone.
- Then adjust the P element.  Keeping the I gain on, adjust `anti_gravity_p_gain` upwards in steps, 50 - 100 - 150.  More P will help control the transient set point error that can arise as I adjusts.  Too much P gain may lead to P oscillation (easy to see in the log).  Some amount of P gain, hopefully close to default, will improve stability during the throttle boost and cut, without adding oscillation.
- For special builds, consider modifying the cutoff or smoothing value, to adjust the timing of the boost effect.  If the quad is heavy and slow, and the pitch wobble with no antiGravity lasts more than about 80 milliseconds, a lower cutoff value will prolong the effect, so try that out.  If the quad is a racer, and the pitch wobble (with no antiGravity) is very brief, the default should be OK, but 6 or 7hz may optimise the timing.  The cut-off is a very sensitive parameter.  Higher cutoff values will allow very fast inputs to generate more gain than slower ones. 
- Note: the cutoff / timing is optimal when the peak of the iTerm boost coincides with the midpoint of the change in iTerm.  You don't want to boost iTerm accumulation rate any other time.

**Changes**
- removal of the old Step mode.
- adds ability to adjust the P contribution (thanks @limonspb)
- uses a PT2 smoothed derivative model with greater overall smoothness, 
- the PT2 cutoff can be modified to customise the timing of the response.
- P boost won't go so high on extremely fast stick travels, reducing the risk of 'too much P' oscillations, 
- less I during the middle of a throttle up
- slightly softer than before for extreme inputs
- smoother behaviour lower speed stick inputs
- Uses an 8 bit value in CLI for antigravity, and changes the scaling factors
- Set 35 there to equal the old 3500.
- updated debugs

The PR was prompted by @limonspb's PR https://github.com/betaflight/betaflight/pull/11666, where he identified significant P oscillations with fast stick inputs.  The old code could push P very high with fast inputs.  Limon's solution was to use limiting and to provide the user with the ability to attenuate (or enhance) the amount of P boost relative to I boost.

We also found significant and previously un-recognised yaw oscillation with standard antigravity boost, due to the higher yaw gains in 4.3.

This PR should solve the problem of excessive boost for really fast movements by a different method.  The PT2 smoothing limits the amount of very fast stick travel that can get through.  It appears to do this quite well.  There is less than 10% difference in the amount of AntiGravity added for a 50ms, 100ms or 150ms full range throttle push or chop.  

There is a big change to the I boost side of antigravity.  Previously it was sustained longer than needed.  In this PR, the I boost now follows exactly same timing as P boost.  This greatly simplifies the code, and logs suggest they share common timing requirements.

**CLI changes:**
- removed the option to change mode, and the old threshold value used by step mode.
- added `anti_gravity_cutoff`, the PT2 cutoff in Hz that smooths the gain function.  Default is 6Hz.  Lower values will smooth out faster movements (making them more equal in max iTerm) and both delay and prolong the effect.  The default of 6Hz is a little bit faster in offset than the old method.
- added `anti_gravity_p_gain`, based on Limon's work, which, when set to 100, returns the default amount of P.  When set to 50, only half the normal P is added.  This allows fine-tuning the relative amounts of P and I in the AntiGravity response.  When set to zero, there is no P boost, like the older iTerm only AntiGravity code.

**Benefits:**
- less likelihood of P oscillation with really fast throttle moves
- better, and adjustable, matching of the timing of the boost effect
- smoother and more predictable boost of P and I during slower throttle inputs

**Possible issues:**
- Default AG Gain may be too high for race builds
- a bit more lag compared to old code, when first pushing throttle up from zero.

**Logging and Debug**
The custom debug is `DEBUG_ANTI_GRAVITY`
Values are * 100.  

- Debug_0: Un-smoothed, simple, raw throttle derivative, will wobble at RC link frequency
- Debug_1: PT2 smoothed throttle derivative, including attenuation at higher throttle positions and when throttling up
- Debug_2: Effective multiplier of user's current iTerm gain by Antigravity.  100 means no gain. 200 means rate of iTerm gain is approximately doubled from normal.
- Debug_3: Multiplier on user's P gain.  100 means adding nothing, 200 means P is twice what it otherwise would be.

Many thanks to @limonspb for identifying the problem.  His PR that laid some of the conceptual foundation for this PR.
